src/screencast/pipewire_screencast.c

```diff
@@ -128,66 +135,87 @@ static const struct pw_stream_events pwr_stream_events = {
+	struct xdpw_state *state = ctx->state;
+
 	pw_loop_enter(state->pw_loop);
 
 	const struct spa_pod *params[1];
 	uint8_t buffer[1024];
 	struct spa_pod_builder b = SPA_POD_BUILDER_INIT(buffer, sizeof(buffer));
 
-	ctx->pwr_context = pw_context_new(state->pw_loop, NULL, 0);
-	if (!ctx->pwr_context) {
-		logprint(ERROR, "pipewire: failed to create context");
-		abort();
-	}
-
-	ctx->core = pw_context_connect(ctx->pwr_context, NULL, 0);
-	if (!ctx->core) {
-		logprint(ERROR, "pipewire: couldn't connect to context");
-		abort();
-	}
-
-	ctx->stream = pw_stream_new(ctx->core, "xdg-desktop-portal-wlr",
+	char name[] = "xdpw-stream-XXXXXX";
+	randname(name + strlen(name) - 6);
+	cast->stream = pw_stream_new(ctx->core, name,
 		pw_properties_new(
 			PW_KEY_MEDIA_CLASS, "Video/Source",
 			NULL));
 
-	if (!ctx->stream) {
+	if (!cast->stream) {
 		logprint(ERROR, "pipewire: failed to create stream");
 		abort();
 	}
-	ctx->stream_state = false;
+	cast->pwr_stream_state = false;
 
 	/* make an event to signal frame ready */
-	ctx->event =
-		pw_loop_add_event(state->pw_loop, pwr_on_event, state);
+	cast->event =
+		pw_loop_add_event(state->pw_loop, pwr_on_event, cast);
+	logprint(TRACE, "pipewire: registered event %p", cast->event);
 
 	params[0] = spa_pod_builder_add_object(&b,
 		SPA_TYPE_OBJECT_Format, SPA_PARAM_EnumFormat,
 		SPA_FORMAT_mediaType,       SPA_POD_Id(SPA_MEDIA_TYPE_video),
 		SPA_FORMAT_mediaSubtype,    SPA_POD_Id(SPA_MEDIA_SUBTYPE_raw),
-		SPA_FORMAT_VIDEO_format,    SPA_POD_Id(pipewire_from_wl_shm(ctx)),
+		SPA_FORMAT_VIDEO_format,    SPA_POD_Id(xdpw_format_pw_from_wl_shm(cast)),
 		SPA_FORMAT_VIDEO_size,      SPA_POD_CHOICE_RANGE_Rectangle(
-			&SPA_RECTANGLE(ctx->simple_frame.width, ctx->simple_frame.height),
+			&SPA_RECTANGLE(cast->simple_frame.width, cast->simple_frame.height),
 			&SPA_RECTANGLE(1, 1),
 			&SPA_RECTANGLE(4096, 4096)),
 		// variable framerate
 		SPA_FORMAT_VIDEO_framerate, SPA_POD_Fraction(&SPA_FRACTION(0, 1)),
 		SPA_FORMAT_VIDEO_maxFramerate, SPA_POD_CHOICE_RANGE_Fraction(
-			&SPA_FRACTION(ctx->framerate, 1),
+			&SPA_FRACTION(cast->framerate, 1),
 			&SPA_FRACTION(1, 1),
-			&SPA_FRACTION(ctx->framerate, 1)));
+			&SPA_FRACTION(cast->framerate, 1)));
 
-	 pw_stream_add_listener (ctx->stream, &ctx->stream_listener,
-	 	&pwr_stream_events, state);
+	pw_stream_add_listener(cast->stream, &cast->stream_listener,
+		&pwr_stream_events, cast);
 
-	pw_stream_connect(ctx->stream,
+	pw_stream_connect(cast->stream,
 		PW_DIRECTION_OUTPUT,
 		PW_ID_ANY,
 		(PW_STREAM_FLAG_DRIVER |
 			PW_STREAM_FLAG_MAP_BUFFERS),
 		params, 1);
 
-	return NULL;
+}
+
+void xdpw_pwr_core_connect(struct xdpw_state *state){
+	struct xdpw_screencast_context *ctx = &state->screencast;
+
+	logprint(TRACE, "pipewire: establishing connection to core");
+
+	if(!ctx->pwr_context){
+		ctx->pwr_context = pw_context_new(state->pw_loop, NULL, 0);
+		if (!ctx->pwr_context) {
+			logprint(ERROR, "pipewire: failed to create context");
+			abort();
```

Comment:
Can we use a return value instead of aborting? Aborting is not great if it's caused by runtime errors.